File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode middleNode(ListNode head) {

    // Slowhead to loop through linkedlist in 1 step
    ListNode slowHead = head;
    // Slowhead to loop through linkedlist in 2 step
    ListNode fastHead = head;

    while(fastHead.next != null && fastHead.next.next != null){
        slowHead = slowHead.next;
        fastHead = fastHead.next.next;
    }

    if(fastHead.next != null){
        slowHead = slowHead.next;
    }

    return slowHead; 
    }
}

File: 0908-middle-of-the-linked-list/0908-middle-of-the-linked-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        # Progress through LinkedList in 1 step
        step_1 = head
        # Progress through LinkedList in 2 step
        step_2 = head

        while step_2.next != None and step_2.next.next != None:
            step_1 = step_1.next
            step_2 = step_2.next.next

        if step_2.next is not None:
            step_1 = step_1.next

        return step_1